File: src/test/java/com/fluxchess/pulse/PieceTest.java
/*
 * Copyright (C) 2013-2016 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */
package com.fluxchess.pulse;

import org.junit.Test;

import static com.fluxchess.pulse.Color.BLACK;
import static com.fluxchess.pulse.Color.WHITE;
import static com.fluxchess.pulse.Piece.*;
import static com.fluxchess.pulse.PieceType.*;
import static com.fluxchess.test.AssertUtil.assertUtilityClassWellDefined;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class PieceTest {

	@Test
	public void testUtilityClass() throws Exception {
		assertUtilityClassWellDefined(Piece.class);
	}

	@Test
	public void testValues() {
		for (int piece : Piece.values) {
			assertThat(Piece.values[piece], is(piece));
		}
	}

	@Test
	public void testValueOf() {
		assertThat(Piece.valueOf(Color.WHITE, PieceType.PAWN), is(Piece.WHITE_PAWN));
		assertThat(Piece.valueOf(Color.WHITE, PieceType.KNIGHT), is(Piece.WHITE_KNIGHT));
		assertThat(Piece.valueOf(Color.WHITE, PieceType.BISHOP), is(Piece.WHITE_BISHOP));
		assertThat(Piece.valueOf(Color.WHITE, PieceType.ROOK), is(Piece.WHITE_ROOK));
		assertThat(Piece.valueOf(Color.WHITE, PieceType.QUEEN), is(Piece.WHITE_QUEEN));
		assertThat(Piece.valueOf(Color.WHITE, PieceType.KING), is(Piece.WHITE_KING));
		assertThat(Piece.valueOf(Color.BLACK, PieceType.PAWN), is(Piece.BLACK_PAWN));
		assertThat(Piece.valueOf(Color.BLACK, PieceType.KNIGHT), is(Piece.BLACK_KNIGHT));
		assertThat(Piece.valueOf(Color.BLACK, PieceType.BISHOP), is(Piece.BLACK_BISHOP));
		assertThat(Piece.valueOf(Color.BLACK, PieceType.ROOK), is(Piece.BLACK_ROOK));
		assertThat(Piece.valueOf(Color.BLACK, PieceType.QUEEN), is(Piece.BLACK_QUEEN));
		assertThat(Piece.valueOf(Color.BLACK, PieceType.KING), is(Piece.BLACK_KING));
	}

	@Test
	public void testGetType() {
		assertThat(Piece.getType(WHITE_PAWN), is(PAWN));
		assertThat(Piece.getType(BLACK_PAWN), is(PAWN));
		assertThat(Piece.getType(WHITE_KNIGHT), is(KNIGHT));
		assertThat(Piece.getType(BLACK_KNIGHT), is(KNIGHT));
		assertThat(Piece.getType(WHITE_BISHOP), is(BISHOP));
		assertThat(Piece.getType(BLACK_BISHOP), is(BISHOP));
		assertThat(Piece.getType(WHITE_ROOK), is(ROOK));
		assertThat(Piece.getType(BLACK_ROOK), is(ROOK));
		assertThat(Piece.getType(WHITE_QUEEN), is(QUEEN));
		assertThat(Piece.getType(BLACK_QUEEN), is(QUEEN));
		assertThat(Piece.getType(WHITE_KING), is(KING));
		assertThat(Piece.getType(BLACK_KING), is(KING));
	}

	@Test
	public void testGetColor() {
		assertThat(Piece.getColor(WHITE_PAWN), is(WHITE));
		assertThat(Piece.getColor(BLACK_PAWN), is(BLACK));
		assertThat(Piece.getColor(WHITE_KNIGHT), is(WHITE));
		assertThat(Piece.getColor(BLACK_KNIGHT), is(BLACK));
		assertThat(Piece.getColor(WHITE_BISHOP), is(WHITE));
		assertThat(Piece.getColor(BLACK_BISHOP), is(BLACK));
		assertThat(Piece.getColor(WHITE_ROOK), is(WHITE));
		assertThat(Piece.getColor(BLACK_ROOK), is(BLACK));
		assertThat(Piece.getColor(WHITE_QUEEN), is(WHITE));
		assertThat(Piece.getColor(BLACK_QUEEN), is(BLACK));
		assertThat(Piece.getColor(WHITE_KING), is(WHITE));
		assertThat(Piece.getColor(BLACK_KING), is(BLACK));
	}
}

File: src/test/cpp/piecetest.cpp
/*
 * Copyright (C) 2013-2019 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

#include "piece.h"
#include "color.h"
#include "piecetype.h"

#include "gtest/gtest.h"

using namespace pulse;

TEST(piecetest, testValues) {
	for (auto piece : piece::values) {
		EXPECT_EQ(piece, piece::values[piece]);
	}
}

TEST(piecetest, testValueOf) {
	EXPECT_EQ(+piece::WHITE_PAWN, piece::valueOf(color::WHITE, PieceType::PAWN));
	EXPECT_EQ(+piece::WHITE_KNIGHT, piece::valueOf(color::WHITE, PieceType::KNIGHT));
	EXPECT_EQ(+piece::WHITE_BISHOP, piece::valueOf(color::WHITE, PieceType::BISHOP));
	EXPECT_EQ(+piece::WHITE_ROOK, piece::valueOf(color::WHITE, PieceType::ROOK));
	EXPECT_EQ(+piece::WHITE_QUEEN, piece::valueOf(color::WHITE, PieceType::QUEEN));
	EXPECT_EQ(+piece::WHITE_KING, piece::valueOf(color::WHITE, PieceType::KING));
	EXPECT_EQ(+piece::BLACK_PAWN, piece::valueOf(color::BLACK, PieceType::PAWN));
	EXPECT_EQ(+piece::BLACK_KNIGHT, piece::valueOf(color::BLACK, PieceType::KNIGHT));
	EXPECT_EQ(+piece::BLACK_BISHOP, piece::valueOf(color::BLACK, PieceType::BISHOP));
	EXPECT_EQ(+piece::BLACK_ROOK, piece::valueOf(color::BLACK, PieceType::ROOK));
	EXPECT_EQ(+piece::BLACK_QUEEN, piece::valueOf(color::BLACK, PieceType::QUEEN));
	EXPECT_EQ(+piece::BLACK_KING, piece::valueOf(color::BLACK, PieceType::KING));
}

TEST(piecetest, testGetType) {
	EXPECT_EQ(+PieceType::PAWN, piece::getType(piece::WHITE_PAWN));
	EXPECT_EQ(+PieceType::PAWN, piece::getType(piece::BLACK_PAWN));
	EXPECT_EQ(+PieceType::KNIGHT, piece::getType(piece::WHITE_KNIGHT));
	EXPECT_EQ(+PieceType::KNIGHT, piece::getType(piece::BLACK_KNIGHT));
	EXPECT_EQ(+PieceType::BISHOP, piece::getType(piece::WHITE_BISHOP));
	EXPECT_EQ(+PieceType::BISHOP, piece::getType(piece::BLACK_BISHOP));
	EXPECT_EQ(+PieceType::ROOK, piece::getType(piece::WHITE_ROOK));
	EXPECT_EQ(+PieceType::ROOK, piece::getType(piece::BLACK_ROOK));
	EXPECT_EQ(+PieceType::QUEEN, piece::getType(piece::WHITE_QUEEN));
	EXPECT_EQ(+PieceType::QUEEN, piece::getType(piece::BLACK_QUEEN));
	EXPECT_EQ(+PieceType::KING, piece::getType(piece::WHITE_KING));
	EXPECT_EQ(+PieceType::KING, piece::getType(piece::BLACK_KING));
}

TEST(piecetest, testGetColor) {
	EXPECT_EQ(+color::WHITE, piece::getColor(piece::WHITE_PAWN));
	EXPECT_EQ(+color::BLACK, piece::getColor(piece::BLACK_PAWN));
	EXPECT_EQ(+color::WHITE, piece::getColor(piece::WHITE_KNIGHT));
	EXPECT_EQ(+color::BLACK, piece::getColor(piece::BLACK_KNIGHT));
	EXPECT_EQ(+color::WHITE, piece::getColor(piece::WHITE_BISHOP));
	EXPECT_EQ(+color::BLACK, piece::getColor(piece::BLACK_BISHOP));
	EXPECT_EQ(+color::WHITE, piece::getColor(piece::WHITE_ROOK));
	EXPECT_EQ(+color::BLACK, piece::getColor(piece::BLACK_ROOK));
	EXPECT_EQ(+color::WHITE, piece::getColor(piece::WHITE_QUEEN));
	EXPECT_EQ(+color::BLACK, piece::getColor(piece::BLACK_QUEEN));
	EXPECT_EQ(+color::WHITE, piece::getColor(piece::WHITE_KING));
	EXPECT_EQ(+color::BLACK, piece::getColor(piece::BLACK_KING));
}
